### ISSUE TITLE
Support TF 1.5 / CUDA 9 / cuDNN 7

### ISSUE THREAD

**Issue by David-Levinthal** [mozilla/DeepSpeech#1194]
For support and discussions, please use our [Discourse forums](https://discourse.mozilla.org/c/deep-speech).

If you've found a bug, or have a feature request, then please create an issue with the following information:

- **Have I written custom code (as opposed to running examples on an unmodified clone of the repository)**:
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
- **TensorFlow installed from (our builds, or upstream TensorFlow)**:
- **TensorFlow version (use command below)**:
- **Python version**: 
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**:
- **GPU model and memory**:
- **Exact command to reproduce**:

You can obtain the TensorFlow version with

```bash
python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"
```

Please describe the problem clearly. Be sure to convey here why it's a bug or a feature request.

Include any logs or source code that would be helpful to diagnose the problem. For larger logs, link to a Gist, not a screenshot. If including tracebacks, please include the full traceback. Try to provide a reproducible test case.
trying to build everything from source.
ubuntu 16.04 with NvidiaV100s. These require cuda9/9.1 and cudnn7, python2.7
gcc (Ubuntu 5.4.0-6ubuntu1~16.04.5) 5.4.0 20160609
bazel version
Build label: 0.9.0
Build target: bazel-out/k8-fastbuild/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Tue Dec 19 09:31:58 2017 (1513675918)
Build timestamp: 1513675918
Build timestamp as int: 1513675918
top of tree TF on 1-22  r1.5rc1
a tf build produces a wheel called
/tmp/tensorflow_pkg/tensorflow-1.5.0rc1-cp27-cp27mu-linux_x86_64.whl

.........
Loading: ^M
Loading: 0 packages loaded^M
ERROR: /home/levinth/tensorflow/native_client/BUILD:18:1: Traceback (most recent call last):
        File "/home/levinth/tensorflow/native_client/BUILD", line 18
                tf_library(name = "deepspeech_model", cpp_cla...", <5 more arguments>)
        File "/home/levinth/tensorflow/tensorflow/compiler/aot/tfcompile.bzl", line 137, in tf_library
                " ".join([(("'" + arg.replace("'", "'\\''... [])])
        File "/home/levinth/tensorflow/tensorflow/compiler/aot/tfcompile.bzl", line 137, in " ".join
                [(("'" + arg.replace("'", "'\\''")) + "'") for arg in (tfcompile_flags or [])]
type 'select' is not iterable^M
ERROR: error loading package 'native_client': Package 'native_client' contains errors^M
INFO: Elapsed time: 0.578s^M
FAILED: Build did NOT complete successfully (2 packages loaded)^M

**Comment by lissyx** [comment]
@David-Levinthal Thanks, but some things:
 - we do not yet have r1.5-based branch
 - please use https://github.com/mozilla/tensorflow and checkout commit 1390dc180e25b5821be80b407ddc5fad73d4ef6a

This should work with CUDA 9 and CuDNN 7 (I did it in the past, and so did reuben: https://github.com/mozilla/DeepSpeech/issues/946#issuecomment-347608611).

Feel free to re-open if you there is a real issue, but please use https://discourse.mozilla.org/c/deep-speech if you just need help.

**Comment by David-Levinthal** [comment]
My concern is that I wanted to use this as part of a HW benchmarking effort.
In particular for silicon being developed by startups.
They (startups) do not have resources to support anything other than top of
tree releases.
So it is important to keep major examples of real codes (ie yours..as
opposed to PTB, assorted CNN benchmarks etc) up to date
please do not close this
see:
https://github.com/David-Levinthal/machine-learning/blob/master/Evaluating%20RNN%20performance%20across%20HW%20platforms.pdf
thanks
d

**Comment by reuben** [comment]
We can always take patches for 1.5 support, but there's no timeline for it at the moment.

**Comment by lissyx** [comment]
@David-Levinthal I'm not so sure how this relates to my reply. First, you are trying to build with the wrong tree of tensorflow (we have a few slight changes that accounts for your tfcompile issue), and with tensorflow version we don't support. Second, I linked you earlier experiments confirming that what we have right now, TensorFlow r1.4 based, works with CUDA 9 and cuDNN7.

**Comment by David-Levinthal** [comment]
understood. Thank you for holding this open.
The issue of having serious applications for evaluating startup hardware is
problematic.
This STT effort is a great component of startup evaluation/support
requirement.
d

**Comment by lissyx** [comment]
@David-Levinthal I still don't understand your problem: you want to be able to run DeepSpeech with TensorFlow r1.5 ? Why isn't it okay with r1.4 + CUDA 9 + cuDNN v7.0 ?

**Comment by David-Levinthal** [comment]
R1.5 will be main release shortly.
I worry about HW startups (ex:Groq, neervana, graphcore,wave
computing,cerebras, tenstorrent) being able to support anything but top of
tree.
These outfits will be trying to bring HW to market this year and need to be
able to test on real code.
as a further example. Tensorflow NMT appears to requires TF r1.5 to
actually run well on Intel processors with MKL/mkldnn
I have been unable to get it to run until r1.5 (OOM).
So it is important that real serious apps (NMT, SOckeye, OpenNMT, KNMT and
STT codes like DeepSpeech) can be used as hurdles.
Does that help give some context?

personally I will start using 1.4 for now..but there is a lot of HW in
development we will all be seeing in the future.
d

**Comment by lissyx** [comment]
@David-Levinthal Yes, thanks! That gives more context. Well, don't worry, we do not plan to keep on r1.4 for ever, but we have several things to complete before starting to work on it :). For example, TF Lite also seems to be r1.5+, and we want to work on that soonish.

**Comment by David-Levinthal** [comment]
FYI
I am HW performance lead for Msft cloud services...so I see a lot of early
HW.
If you look at that RNN paper on the github site you'll see why I am so
concerned about having serious applications for holding the HW vendors'
feet to the fire with.
That was a clumsy sentence
:-)

**Comment by lissyx** [comment]
@David-Levinthal Nice! Well, if you require TF r1.5-based, and if you have time to hack, @reuben fixed #1197, but you can already try to build everything, I would just advise to disable `tf_library` statement in our `native_client/BUILD` file, since this is what triggered your error at first.

**Comment by reuben** [comment]
FWIW, now that #1197 landed we are source compatible with TF 1.5 modulo the AOT stuff, so if you remove all references to tfcompile, deepspeech_model, if_native_model, etc. from native_client/BUILD you should be able to build and test with TF 1.5.

**Comment by reuben** [comment]
Er, I should have refreshed the page first, @lissyx already covered this :)

**Comment by David-Levinthal** [comment]
thanks...I really appreciate your understanding of the importance.
d

**Comment by David-Levinthal** [comment]
just to check things..I have attempted to build this from source using TF
r1.4 built from source after git checkout r1.4 to force the version  (new
HW vendors are going to have to do everything from source)
using cuda 9.1 cudnn 7.0.4
and setting PYTHONPATH to the directory where I installed the R1.4 TF
wheel   (I do this as I am maintaining several TF versions)
requirements was editted to remove the tf r1.4 line
I move the TF source directory that I used to build TF and clone another
download of TF, checkout r1.4, run configure, link the native_client into
~/source tensorflow
invoke
bazel build -c opt --copt=-O3 //tensorflow:libtensorflow_cc.so
//tensorflow:libtensorflow_framework.so //native_client:deepspeech
//native_client:deepspeech_utils
//native_client:libctc_decoder_with_kenlm.so //native_client:generate_trie
> native_build2.log 2>&1
and get
ERROR: /home/levinth/tensorflow/native_client/BUILD:18:1: Traceback (most
recent call last):
        File "/home/levinth/tensorflow/native_client/BUILD", line 18
                tf_library(name = "deepspeech_model", cpp_cla...", <5 more
arguments>)
        File
"/home/levinth/tensorflow/tensorflow/compiler/aot/tfcompile.bzl", line 171,
in tf_library
                tfcompile_flags.find("--gen_program_shape")
type 'select' has no method find(string)^M
ERROR: error loading package 'native_client': Package 'native_client'
contains errors^M

I am clearly screwing up..but can you throw me a clue  :-)

On Wed, Jan 24, 2018 at 8:43 AM, David Levinthal <david.levinthal1@gmail.com

**Comment by lissyx** [comment]
@David-Levinthal Yes, this is what we said above: you need to remove the use of `tf_library` and matching `if_native` uses.

**Comment by David-Levinthal** [comment]
that was what I suspected...but was trying to stick as close to the readme
instructions on the github site first
BTW. I have raised the idea of focusing on the Mozilla deepspeech (instead
of the old Ford distro) with some of the HW teams.

**Comment by lissyx** [comment]
@David-Levinthal Since this is derailing from pure TensorFlow r1.5 support, feel free to open a thread on https://discourse.mozilla.org/c/deep-speech if you need more help / want to discuss things :).

**Comment by David-Levinthal** [comment]
https://github.com/tensorflow/tensorflow/issues/16412
issue on build from source documentation needing a bit of clarification
:-)

**Comment by lissyx** [comment]
@David-Levinthal Why did you filed that against upstream TensorFlow ?

**Comment by David-Levinthal** [comment]
I did???...yarghhhh
:-(

**Comment by David-Levinthal** [comment]
https://github.com/mozilla/DeepSpeech/issues/1199

**Comment by lissyx** [comment]
@David-Levinthal You should be able to play with some binaries from https://tools.taskcluster.net/index/project.deepspeech.deepspeech.native_client.tf-master, it's CUDA 9.0, CuDNN v7 and based on current master of tensorflow.

**Comment by David-Levinthal** [comment]
BTW..just to ask a question..this is with TF1.4/cuda 9/cudnn7.1
I added a timer that prints out the time used for every 100 increments of
current_step during training..but even after 18 epoch the default printout
of the loss is inf?
current_step = 79600 time_delta = 166.307541
current_step = 79700 time_delta = 176.724349
current_step = 79800 time_delta = 192.317842
I Training of Epoch 18 - loss: inf
current_step = 79900 time_delta = 176.476619
current_step = 80000 time_delta = 46.461117
current_step = 80100 time_delta = 52.504259

any idea what am I getting wrong?python -u DeepSpeech.py --train_files
../Common_voice/cv-valid-train.csv,../Common_voice/cv-other-train.csv
--dev_files ../Common_voice/cv-valid-dev.csv --test_files
../Common_voice/cv-valid-test.csv --train_batch_size 20 --dev_batch_size 1
--test_batch_size 1 --n_hidden 512 --epoch 50 --checkpoint_dir
/home/levinth/DS_checkpoint --summary_dir ../DS_summary --summary_secs 10 >
deepspeech20_512.log 2>&1
Should I submit another bug..or is this trivially obvious and i don't see
it?

**Comment by David-Levinthal** [comment]
Fantastic. Thank you!

**Comment by David-Levinthal** [comment]
Thank you for the direction...
you're team is great!
:-)
d

**Comment by David-Levinthal** [comment]
that did not fix it..added more debug code around the session.run call
it seems the session run call occasionally returns inf
also I noticed that the variable total_loss gets zeroed before each call to
session run (line 1615 0f Deepspeech.py)
as the loop at 1635 always has tripcount 1 in my runs
for job_step in range(job.steps):
ie job_steps is always 1

python -u DeepSpeech.py --train_files
../Common_voice/cv-valid-train.csv,../Common_voice/cv-other-train.csv
--dev_files ../Common_voice/cv-valid-dev.csv --test_files
../Common_voice/cv-valid-test.csv --train_batch_size 12 --dev_batch_size 1
--test_batch_size 1 --n_hidden 512 --epoch 50 --checkpoint_dir
/home/levinth/DS_checkpoint --summary_dir ../DS_summary --summary_secs 10
--learning_rate 0.0001 --wer_log_pattern "GLOBAL LOG:
logwer('${COMPUTE_ID}', '%s', '%s', %f)"--log_level 0  --display_step 1 >
deepspeech20_512_2.log 2>&1
I STARTING Optimization
 current_step = 1, batch_loss = 100.752945
 current_step = 1, total_loss = 100.752945 total_mean_edit_distance
0.992792
 current_step = 2, batch_loss = 104.521294
 current_step = 2, total_loss = 104.521294 total_mean_edit_distance
0.858542
 current_step = 3, batch_loss = 67.393005
 current_step = 3, total_loss = 67.393005 total_mean_edit_distance
0.793802
 current_step = 4, batch_loss = inf
 current_step = 4, total_loss = inf total_mean_edit_distance  0.832232
 current_step = 5, batch_loss = 91.163597
 current_step = 5, total_loss = 91.163597 total_mean_edit_distance
0.772215
 current_step = 6, batch_loss = 88.259605
 current_step = 6, total_loss = 88.259605 total_mean_edit_distance
0.823895
 current_step = 7, batch_loss = inf
 current_step = 7, total_loss = inf total_mean_edit_distance  0.804266
 current_step = 8, batch_loss = 74.751190
 current_step = 8, total_loss = 74.751190 total_mean_edit_distance
0.837069




On Mon, Feb 12, 2018 at 8:40 AM, David Levinthal <david.levinthal1@gmail.com

**Comment by kdavis-mozilla** [comment]
@David-Levinthal You've changed the network architecture
```
...
  --n_hidden 512 
...
```
too so our hyperparameters will not necessarily work for your new architecture.
